Eviction purger

aweisberg opened the pull request.

belliottsmith committed:
eviction

belliottsmith committed:
Refactor: apply truncation on access, rather than in bulk; necessary for efficient C* integration.
This patch has a lot of changes to support this and efficient compaction.

belliottsmith committed:
fixing tests

belliottsmith committed:
refactor: introduce Participants to model keys we know to be visible to other transactions (i.e. excluding any hidden home key)

belliottsmith committed:
improvements to NotifyWaitingOn

iamaleksey committed:
Add missing Command.SerializationSupport methods

iamaleksey committed:
Add missing equals() impls to WaitingOn and SimpleBitSet

belliottsmith committed:
feedback

belliottsmith committed:
Update accord-core/src/main/java/accord/messages/CheckStatus.java

Co-authored-by: dcapwell <dcapwell@gmail.com>

belliottsmith committed:
feedback

belliottsmith committed:
feedback

belliottsmith committed:
test

belliottsmith committed:
remove progressShard from Command

belliottsmith committed:
remove unused imports

belliottsmith committed:
standardise on ShardDurable not LocallyDurable

iamaleksey committed:
fixup

iamaleksey committed:
Add serialization support for SyncPoint

iamaleksey committed:
nits

belliottsmith committed:
add shouldTruncate variant

aweisberg committed:
Fix bounds checking in ReducingIntervalMap

aweisberg committed:
Needs review - Tolerate empty epochs fix

aweisberg committed:
Needs review - fix for AccordCQLTest.testMultipleShards where not enough epochs end up being included

belliottsmith committed:
refactor/centralise invalidation inference

belliottsmith committed:
Clarify/stream-line state machine enums

belliottsmith committed:
add SimpleBitSetTest

belliottsmith committed:
Clarify/stream-line state machine enums

belliottsmith committed:
Clarify/stream-line state machine enums

belliottsmith committed:
Submit executeAt to WaitUntilApplied in case remote side is Truncated

belliottsmith committed:
Clarify/stream-line state machine enums

belliottsmith committed:
Clarify/stream-line state machine enums

belliottsmith committed:
fix c* integration

aweisberg committed:
Fixes to get C* compiling and passing style checks

aweisberg committed:
Get AccordCQLTest passing

aweisberg committed:
Fix initialization exception

aweisberg committed:
Fix TrackerReconcilerTest creating configurations with nodes with no shards triggering assertion in TopologyUtils.initialTopology. RF is not a factor in calculating how many nodes are possible because the placement algorithm uses each node RF times if possible.

aweisberg committed:
Fix bad rebease/merge of TCM/Bootstrap changes to TopologyManagerTest

belliottsmith committed:
add neater Range and Ranges toString

belliottsmith committed:
add neater Range and Ranges toString

belliottsmith committed:
revert incorrect fix for topology calculation

belliottsmith committed:
notify self and ensure topology manager is synchronized

belliottsmith committed:
revert bad fix

belliottsmith committed:
feedback

belliottsmith committed:
fix quick fix to feedback

belliottsmith committed:
adopt David's feedback

belliottsmith committed:
fix condition to mark waiting on apply

belliottsmith committed:
evict CommandsForKey timestamps when they're old

belliottsmith committed:
permit clearing lastExecutedMicros, and rename to lastExecutedHlc so as not to leak C* details

belliottsmith committed:
handle range removals across epochs, and fix TopologyRandomizer (lots of new breakages to address)

belliottsmith committed:
FetchCoordinator sends missing apply

belliottsmith committed:
address some issues and improve docs

belliottsmith committed:
wip: address some issues discovered with re-enablement of topology changes

aweisberg committed:
Squash durability scheduling

aweisberg committed:
Changes to support AccordPurger

aweisberg committed:
Fix AbstractKeys.toRoute incorrectly not being able to determine if the home key participates because the RoutingKey type and Key don't interoperate correctly with binary search

aweisberg committed:
Fix initialization issues and mutation safety in CommonAttributes and Command

belliottsmith committed:
draft: coordinate shard durability

aweisberg committed:
Remove unused LocalAndGlobal in TopologyManager

aweisberg committed:
Review feedback starting at https://github.com/apache/cassandra-accord/pull/50#pullrequestreview-1535467280

dcapwell commented on accord-core/src/test/java/accord/impl/AbstractConfigurationServiceTest.java:
```diff
@@ -43,6 +45,7 @@
 
 public class AbstractConfigurationServiceTest
 {
+    // C* uses this, so we can't use the jupiter assertions
```
does this mean C* excludes the junit 5 dependency?  If so, we might want to rethink this module setup to make this more explicit

aweisberg replied:
Yes C* uses it, it was added as part of TCM/Bootstrap. This commit already merged upstream in `trunk` and I am not sure why the PR is showing it? It should end up being an a no-op when merged.

dcapwell commented on accord-core/src/main/java/accord/impl/InMemoryCommandStore.java:
```diff
@@ -459,45 +408,63 @@ void update(Ranges add)
     }
 
-    class CFKLoader implements CommandLoader<TxnId>
+    static class CFKEntry extends TxnId
```
ill split this into a different PR once evict is in, but to make sure we actually test the cache-miss case I added a LRU~ish "cache" and dropped this class as its broken... we are pulling a globally modifiable reference within an immutable object (CommandsForKeys)?

dcapwell commented on accord-core/src/main/java/accord/local/CommonAttributes.java:
```diff
@@ -84,6 +80,7 @@ public TxnId txnId()
 
         public Mutable txnId(TxnId txnId)
         {
+            Invariants.checkState(this != EMPTY_ATTRS && this != EMPTY_ATTRS_NONE_PARTIAL_DEPS, "Don't mutate the global empty instances");
```
in this case I kinda feel extending `CommonAttributes.Mutable` might be best... though its more code


```
private static class ImmutableAttributes extends CommonAttributes.Mutable
{
    priviate final String name; // to know if this was empty or empty w/ deps.none
```

aweisberg replied:
I'll remove the EMPTY instances so it won't matter

dcapwell commented on accord-core/src/main/java/accord/local/Command.java:
```diff
@@ -685,25 +819,29 @@ public Ballot accepted()
     {
-        private final ImmutableSortedSet<TxnId> waitingOnCommit;
-        private final ImmutableSortedMap<Timestamp, TxnId> waitingOnApply;
+        public static final Committed EMPTY = new Committed(SaveStatus.Committed);
```
spoke to you about this in Slack, but don't think we should have this.  This is a C* concern for memory size tracking, but as this updates the public API for accord it means that this can't be changed once a release happens (as it would be a breaking change).

I feel C* should handle this by calling the constructor, or SerializerSupport in C* code, as this is only a concern there

aweisberg replied:
I'll remove them

aweisberg replied:
Why is it a breaking change? The empty object returned is always measurable at runtime so if it changes so will the empty size. Empty size is for in memory footprint, serialization explicitly goes through the fields.

dcapwell commented on accord-core/src/main/java/accord/primitives/AbstractKeys.java:
```diff
@@ -229,42 +236,64 @@ public final long foldl(IndexedFoldToLong<? super K> fold, long param, long init
-            if (resultCount == 0 || !withKey.equals(result[resultCount - 1]))
-                result[resultCount++] = withKey;
-            resultCount += copyToRoutingKeys(keys, insertPos, result, resultCount, keys.length - insertPos);
+            // TODO review this is a pretty terrible work around for the fact that you can't binarySearch these
```
```suggestion
            // TODO (now): review this is a pretty terrible work around for the fact that you can't binarySearch these
```

this project uses a custom TODO syntax that tags comments... I have a script to let me keep track of these tags...

```
$ ./remaming-todos.sh | head
## efficiency: 95
| TODO (required, efficiency): callbacks should be associated with a commandStore for processing to avoid this                  | modules/accord/accord-core/src/main/java/accord/impl/SimpleProgressLog.java              |
| Set<Id> notAwareOfDurability; // TODO (easy, efficiency): use Agrona's IntHashSet as soon as Node.Id switches from long to... | modules/accord/accord-core/src/main/java/accord/impl/SimpleProgressLog.java              |
| Set<Id> notPersisted;         // TODO (easy, efficiency): use Agrona's IntHashSet as soon as Node.Id switches from long to... | modules/accord/accord-core/src/main/java/accord/impl/SimpleProgressLog.java              |
| TODO (desirable, efficiency): we can probably simplify things by requiring (empty) Apply messages to be sent also to the c... | modules/accord/accord-core/src/main/java/accord/impl/SimpleProgressLog.java              |
| TODO (desirable, efficiency): forward to local progress shard for processing (if known)                                       | modules/accord/accord-core/src/main/java/accord/impl/SimpleProgressLog.java              |
| TODO (desirable, efficiency): if we are co-located with the home shard, don't need to do anything unless we're in a           | modules/accord/accord-core/src/main/java/accord/impl/SimpleProgressLog.java              |
| TODO (find library, efficiency): this is obviously super inefficient, need some range map                                     | modules/accord/accord-core/src/main/java/accord/impl/InMemoryCommandStore.java           |
| TODO (desired, efficiency): this can be made more efficient by batching by epoch                                              | modules/accord/accord-core/src/main/java/accord/impl/InMemoryCommandStore.java           |
| TODO (find lib, efficiency): this is super inefficient, need to store Command in something queryable                          | modules/accord/accord-core/src/main/java/accord/impl/InMemoryCommandStore.java           |
$ ./remaming-todos.sh -m tags
  95 efficiency
  66 low priority
  53 desired
  52 expected
  39 required
  25 testing
  24 consider
  14 easy
  13 now
  13 clarity
   8 desirable
   4 safety
   3 correctness
   3 API
   2 semantics
   2 liveness
   2 exceptions
   2 cleanup
   2 API consistency
   2 ?
   1 validation
   1 urgent
   1 topology
   1 soon
   1 simple
   1 progress
   1 performance
   1 later
   1 find library
   1 find lib
   1 exepected
   1 error handling
   1 behaviour
```

aweisberg replied:
I can remove the TODO, Benedict already took a look at it. I'll start using that syntax, thanks.

dcapwell commented on accord-core/src/main/java/accord/primitives/AbstractKeys.java:
```diff
@@ -229,42 +236,64 @@ public final long foldl(IndexedFoldToLong<? super K> fold, long param, long init
+            // two incompatible types
+            RoutingKey[] result = new RoutingKey[keys.length];
+            for (int i = 0; i < keys.length; i++)
+            {
```
style: `{}` not needed, can remove

dcapwell reviewed (approved):
+1 to the changes on-top of eviction, leaving that patch out of review of this as thats a different PR

belliottsmith committed:
comments

aweisberg committed:
Purger review feedback

belliottsmith committed:
fix

aweisberg committed:
Remove empty Commands

dcapwell committed:
Feedback

aweisberg committed:
Fix imports

aweisberg committed:
Fix method signature so it works in C* in intellij

aweisberg committed:
Add getBlockingAndRethrow to AsyncChains

dcapwell committed:
feedback for pr 3540

aweisberg committed:
Make Commands.Cleanup.appliesIfNot public so it can be used outside Accord

dcapwell committed:
revert back to have an accumulating reducing async chain

dcapwell committed:
checkpointing for the night

belliottsmith committed:
enable durability

aweisberg committed:
Add back removed import from Cluster

aweisberg committed:
Fix TopologyRandomizer being unable to generate mutations for a single shard/range topology

Random.nextInt has an exclusive upper bound and it was being passed current.size() which is 1 so the only value that could be generated was 0 resulting in a null result which the caller then loops infinitely.

Reviewed by David

aweisberg commented on accord-core/src/main/java/accord/local/CommandStores.java:
```diff
@@ -351,6 +350,7 @@ private synchronized TopologyUpdate updateTopology(Node node, Snapshot prev, Top
             Ranges removeRanges = subtracted.slice(current, Minimal);
             if (!removeRanges.isEmpty())
             {
+                // TODO (required): This is updating the a non-volatile field in the previous Snapshot, why modify it at all, even with volatile the guaranteed visibility is weak even with mutual exclusion
```
@belliottsmith Why does this update the previous snapshot in place? We can't guarantee anything in flight reads it so they might as well read it out of the current snapshot?

Can we make a copy to avoid the mutability issue?

belliottsmith replied:
I'm not sure what you mean? This is the snapshot we're updating, inside the snapshot holder. This is safe published because all of the fields are final.

dcapwell committed:
backport more gens

dcapwell committed:
backport runs

dcapwell committed:
feedback

dcapwell committed:
TODO comment showing / 0 issue

dcapwell committed:
fixed a bug where timeMillis is not setup correctly

dcapwell committed:
feedback: use setFailure for schedule as only one callpath can set

dcapwell committed:
feedback: use setFailure as it does the same thing

dcapwell committed:
feedback

dcapwell committed:
docs

dcapwell committed:
feedback: get cacnel to actually work

dcapwell committed:
feedback: can do a faster select when first/second are provided

dcapwell committed:
if the object.toString fails, document this in the values list

belliottsmith committed:
fix fix

belliottsmith committed:
Fix DurableBefore.global, and usages of RedundantBefore

dcapwell committed:
use a useful error msg

dcapwell committed:
pick of set now documents why we sort

dcapwell committed:
allow gens that extend range and key rather than forcing a cast

belliottsmith committed:
tweak frequency of durability

dcapwell committed:
feedback: dont use default impl

dcapwell committed:
document why we sort and special case NavigableSet as the set is already sorted

aweisberg committed:
Import fixes
